File: sdks/ruby/lib/backlex/filter.rb
# frozen_string_literal: true

module Backlex
  # Static condition constructors — a Ruby port of the leaf/logical helpers in
  # query.ts. Compose them and pass to QueryBuilder#where. Everything compiles to
  # the canonical JSON Condition the REST API speaks.
  #
  #   rows = client.from("orders").query
  #     .where(Backlex::Filter.and_(
  #       Backlex::Filter.eq("status", "active"),
  #       Backlex::Filter.gte("total", 100),
  #       Backlex::Filter.rel("customer", Backlex::Filter.eq("tier", "gold")), # -> "customer.tier"
  #       Backlex::Filter.gte("placed_at", Backlex::Filter.now(sub: { "months" => 1 })),
  #     ))
  #     .select("id", "total", "customer.name")
  #     .order_by("-placed_at", "id")
  #     .limit(50)
  #     .list
  module Filter
    module_function

    def leaf(field, op, value)
      { field => { op => value } }
    end

    def eq(f, v); leaf(f, "_eq", v); end
    def neq(f, v); leaf(f, "_neq", v); end
    def gt(f, v); leaf(f, "_gt", v); end
    def gte(f, v); leaf(f, "_gte", v); end
    def lt(f, v); leaf(f, "_lt", v); end
    def lte(f, v); leaf(f, "_lte", v); end
    def in_(f, vs); leaf(f, "_in", vs); end
    def nin(f, vs); leaf(f, "_nin", vs); end
    def between(f, lo, hi); leaf(f, "_between", [lo, hi]); end
    def is_null(f, is_null = true); leaf(f, "_null", is_null); end
    def empty(f); leaf(f, "_empty", true); end
    def nempty(f); leaf(f, "_nempty", true); end
    def contains(f, v); leaf(f, "_contains", v); end
    def icontains(f, v); leaf(f, "_icontains", v); end
    def starts_with(f, v); leaf(f, "_starts_with", v); end
    def ends_with(f, v); leaf(f, "_ends_with", v); end

    def and_(*conds); { "$and" => conds }; end
    def or_(*conds); { "$or" => conds }; end
    def not_(cond); { "$not" => cond }; end

    # Traverse a relation one hop: every leaf key produced by +conds+ is prefixed
    # with "head.". Multiple conds are ANDed first.
    def rel(head, *conds)
      inner = conds.length == 1 ? conds[0] : { "$and" => conds }
      prefix_keys(inner, head)
    end

    # Relative-date value, e.g. Filter.now(sub: { "months" => 1 }).
    def now(add: nil, sub: nil)
      opts = {}
      opts["add"] = add if add
      opts["sub"] = sub if sub
      { "$now" => opts }
    end

    def prefix_keys(cond, head)
      return { "$and" => cond["$and"].map { |c| prefix_keys(c, head) } } if cond["$and"].is_a?(Array)
      return { "$or" => cond["$or"].map { |c| prefix_keys(c, head) } } if cond["$or"].is_a?(Array)
      return { "$not" => prefix_keys(cond["$not"], head) } if cond["$not"].is_a?(Hash)

      out = {}
      cond.each { |k, v| out["#{head}.#{k}"] = v }
      out
    end

    # Turn any accepted filter shape into the canonical Condition: handles
    # $and/$or/$not (and their _ aliases) and implicit equality
    # ({ "status" => "active" } -> { "status" => { "_eq" => "active" } }). Idempotent.
    def normalize(raw)
      return {} unless raw.is_a?(Hash)

      a = raw["$and"] || raw["_and"]
      return { "$and" => a.map { |c| normalize(c) } } if a.is_a?(Array)

      o = raw["$or"] || raw["_or"]
      return { "$or" => o.map { |c| normalize(c) } } if o.is_a?(Array)

      if raw.key?("$not") || raw.key?("_not")
        return { "$not" => normalize(raw["$not"] || raw["_not"]) }
      end

      out = {}
      raw.each do |k, v|
        out[k] =
          if v.is_a?(Hash) && comparison?(v) then v
          elsif v.is_a?(Hash) then v # unknown object shape — pass through
          else { "_eq" => v }
          end
      end
      out
    end

    def comparison?(hash)
      !hash.empty? && hash.keys.all? { |k| k.to_s.start_with?("_") }
    end
  end
end

File: sdks/rust/src/filter.rs
//! Static condition constructors — a Rust port of the leaf/logical helpers in
//! query.ts. Compose them and pass to [`crate::QueryBuilder::filter`]. Everything
//! compiles to the canonical JSON `Condition` (a [`serde_json::Value`] object) the
//! REST API speaks. This module is pure (no IO) and compiles to `wasm32`.
//!
//! ```ignore
//! use backlex::filter as f;
//! let cond = f::and(vec![
//!     f::eq("status", "active".into()),
//!     f::gte("total", 100.into()),
//!     f::rel("customer", vec![f::eq("tier", "gold".into())]), // -> "customer.tier"
//!     f::gte("placed_at", f::now(None, Some([("months".into(), 1.into())].into()))),
//! ]);
//! ```

use serde_json::{Map, Value};

fn obj(key: &str, value: Value) -> Value {
    let mut m = Map::new();
    m.insert(key.to_string(), value);
    Value::Object(m)
}

fn leaf(field: &str, op: &str, value: Value) -> Value {
    obj(field, obj(op, value))
}

pub fn eq(f: &str, v: Value) -> Value { leaf(f, "_eq", v) }
pub fn neq(f: &str, v: Value) -> Value { leaf(f, "_neq", v) }
pub fn gt(f: &str, v: Value) -> Value { leaf(f, "_gt", v) }
pub fn gte(f: &str, v: Value) -> Value { leaf(f, "_gte", v) }
pub fn lt(f: &str, v: Value) -> Value { leaf(f, "_lt", v) }
pub fn lte(f: &str, v: Value) -> Value { leaf(f, "_lte", v) }
pub fn in_(f: &str, vs: Vec<Value>) -> Value { leaf(f, "_in", Value::Array(vs)) }
pub fn nin(f: &str, vs: Vec<Value>) -> Value { leaf(f, "_nin", Value::Array(vs)) }
pub fn between(f: &str, lo: Value, hi: Value) -> Value { leaf(f, "_between", Value::Array(vec![lo, hi])) }
pub fn is_null(f: &str, is_null: bool) -> Value { leaf(f, "_null", Value::Bool(is_null)) }
pub fn empty(f: &str) -> Value { leaf(f, "_empty", Value::Bool(true)) }
pub fn nempty(f: &str) -> Value { leaf(f, "_nempty", Value::Bool(true)) }
pub fn contains(f: &str, v: &str) -> Value { leaf(f, "_contains", Value::String(v.into())) }
pub fn icontains(f: &str, v: &str) -> Value { leaf(f, "_icontains", Value::String(v.into())) }
pub fn starts_with(f: &str, v: &str) -> Value { leaf(f, "_starts_with", Value::String(v.into())) }
pub fn ends_with(f: &str, v: &str) -> Value { leaf(f, "_ends_with", Value::String(v.into())) }

pub fn and(conds: Vec<Value>) -> Value { obj("$and", Value::Array(conds)) }
pub fn or(conds: Vec<Value>) -> Value { obj("$or", Value::Array(conds)) }
pub fn not(cond: Value) -> Value { obj("$not", cond) }

/// Traverse a relation one hop: every leaf key produced by `conds` is prefixed
/// with `head.`. Multiple conds are ANDed first.
pub fn rel(head: &str, conds: Vec<Value>) -> Value {
    let inner = if conds.len() == 1 {
        conds.into_iter().next().unwrap()
    } else {
        obj("$and", Value::Array(conds))
    };
    prefix_keys(&inner, head)
}

/// Relative-date value, e.g. `now(None, Some(map))`.
pub fn now(add: Option<Map<String, Value>>, sub: Option<Map<String, Value>>) -> Value {
    let mut opts = Map::new();
    if let Some(a) = add {
        opts.insert("add".into(), Value::Object(a));
    }
    if let Some(s) = sub {
        opts.insert("sub".into(), Value::Object(s));
    }
    obj("$now", Value::Object(opts))
}

fn prefix_keys(cond: &Value, head: &str) -> Value {
    let o = match cond.as_object() {
        Some(o) => o,
        None => return cond.clone(),
    };
    if let Some(arr) = o.get("$and").and_then(|v| v.as_array()) {
        return obj("$and", Value::Array(arr.iter().map(|c| prefix_keys(c, head)).collect()));
    }
    if let Some(arr) = o.get("$or").and_then(|v| v.as_array()) {
        return obj("$or", Value::Array(arr.iter().map(|c| prefix_keys(c, head)).collect()));
    }
    if let Some(n) = o.get("$not") {
        if n.is_object() {
            return obj("$not", prefix_keys(n, head));
        }
    }
    let mut out = Map::new();
    for (k, v) in o {
        out.insert(format!("{}.{}", head, k), v.clone());
    }
    Value::Object(out)
}

/// Turn any accepted filter shape into the canonical Condition: handles
/// $and/$or/$not (and their `_` aliases) and implicit equality
/// (`{"status":"active"}` -> `{"status":{"_eq":"active"}}`). Idempotent.
pub fn normalize(raw: &Value) -> Value {
    let o = match raw.as_object() {
        Some(o) => o,
        None => return Value::Object(Map::new()),
    };

    if let Some(arr) = o.get("$and").or_else(|| o.get("_and")).and_then(|v| v.as_array()) {
        return obj("$and", Value::Array(arr.iter().map(normalize).collect()));
    }
    if let Some(arr) = o.get("$or").or_else(|| o.get("_or")).and_then(|v| v.as_array()) {
        return obj("$or", Value::Array(arr.iter().map(normalize).collect()));
    }
    if o.contains_key("$not") || o.contains_key("_not") {
        let inner = o.get("$not").or_else(|| o.get("_not")).unwrap();
        return obj("$not", normalize(inner));
    }

    let mut out = Map::new();
    for (k, v) in o {
        // Any object (a comparison map or nested shape) passes through unchanged;
        // a scalar or array value gets implicit equality.
        let val = if v.is_object() {
            v.clone()
        } else {
            obj("_eq", v.clone())
        };
        out.insert(k.clone(), val);
    }
    Value::Object(out)
}
